### DISCUSSION TITLE
how to update the type of session in session callback?

### DISCUSSION THREAD

**Discussion by bhatvikrant** [nextauthjs/next-auth#2979] [Help] [answered]
I am using typescript and my [...next-auth].tsx file looks like this: 

```ts
import NextAuth, { Awaitable, Session, User } from "next-auth";
// import GithubProvider from "next-auth/providers/github";
import GoogleProvider from "next-auth/providers/google";

type ExtendedUserType = User & { username?: string; uid?: string };

export default NextAuth({
  // Configure one or more authentication providers
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
    // ...add more providers here
  ],
  pages: {
    signIn: "/auth/signin", // for custom sign in page
  },
  callbacks: {
    async session({ session, token, user }): Awaitable<Session> {
      (session.user as ExtendedUserType).username = session.user?.name
        ?.split(" ")
        .join("")
        .toLocaleLowerCase();

      (session.user as ExtendedUserType).uid = token.sub;

      return session;
    },
  },
});


```

when fetching the session using the `useSession` hook I do not get Typescript autocompletion for the newly added keys - 'username', 'uid'

<img width="568" alt="Screenshot 2021-10-17 at 2 17 01 PM" src="https://user-images.githubusercontent.com/50735025/137619419-c873d7a8-9c97-4815-a373-ae68b58a7f95.png">

I am also getting this typescript error for `Awaitable<Session>` type
<img width="1117" alt="Screenshot 2021-10-17 at 2 18 25 PM" src="https://user-images.githubusercontent.com/50735025/137619531-9ecc894c-fd38-40e2-8e5d-e1bd847b2e0b.png">

**Comment by balazsorban44** [comment]
https://next-auth.js.org/getting-started/typescript#module-augmentation

**Reply by bhatvikrant** [reply]
Thankyou! This works! But how can I extend the Session interface? so that I only have to add the new custom properties and not the already existing ones? Thankyou once again!

**Comment by deadcoder0904** [comment, marked as answer]
Add the following `types/**/*.ts` in `tsconfig.json`:

```json
{ 
  .
  .
  .
  "include": ["next-env.d.ts", "types/**/*.ts", "**/*.ts", "**/*.tsx"],
  .
  .
  .
}
```

Then add the following for `session.user`.

# types/next-auth.d.ts

```ts
import NextAuth from "next-auth"

declare module "next-auth" {
  /**
   * Returned by `useSession`, `getSession` and received as a prop on the `SessionProvider` React Context
   */
  interface Session {
    user: {
      /** The user's name. */
      name: string
    }
  }
}
```

**Reply by bhatvikrant** [reply]
Thankyou! This works! But how can I extend the Session interface? so that I only have to add the new custom properties and not the already existing ones? Thankyou once again!

**Reply by deadcoder0904** [reply]
@bhatvikrant afaik `Session` interface is this one only. `interface`'s merge existing properties together so when you specify `user.name` like above, it'll merge with all the other existing properties.

Maybe try different stuff till TS doesn't yell or be more precise with your question so I can answer :)

**Reply by balazsorban44** [reply]
Module augmentation works by "merging" the already existing interface with your overrides, so you won't have to redefine already existing properties. Read more about it here: https://next-auth.js.org/getting-started/typescript#module-augmentation

**Reply by bhatvikrant** [reply]
This is how my `next-auth.d.ts` looks like
```js
import NextAuth from "next-auth";

declare module "next-auth" {
  /**
   * Returned by `useSession`, `getSession` and received as a prop on the `SessionProvider` React Context
   */
  interface Session {
    user: {
      // name: string;
      // email: string;
      // image: string;

      // Custom fields
      uid: string;
      username: string;
    };
  }
}

```

but when I am using the useSession hook, I only get the TS competition for the properties that I added

<img width="542" alt="Screenshot 2021-10-18 at 4 09 19 PM" src="https://user-images.githubusercontent.com/50735025/137715766-64d9551d-fdc6-48e1-bd7b-cf2d87e60ada.png">

My next-auth version is: 
```js
"next-auth": "4.0.0-beta.4"
```

**Reply by balazsorban44** [reply]
:shrug:. Probably adding three extra type fields isn't that bad, and more explicit anyway. Think about the next person coming to look at your code, they wouldn't know where name, email and image would come from.

**Reply by bhatvikrant** [reply]
Yes 😂.  that is what I am currently doing, I just wanted to know if there is any other way. Thankyou though.

**Reply by deadcoder0904** [reply]
@bhatvikrant weird, works for me with `account.` by adding only extra 2 properties:

<img width="1153" alt="Screenshot 2021-10-18 at 6 49 24 PM" src="https://user-images.githubusercontent.com/16436270/137739132-85915bb4-99c8-4280-bbd4-7fe4a11075ab.png">

And my `Account` interface only has `oauth_token` & `oauth_token_secret`:

```ts
declare module 'next-auth' {
	/**
	 * Usually contains information about the provider being used
	 * and also extends `TokenSet`, which is different tokens returned by OAuth Providers.
	 */
	interface Account {
		oauth_token: string
		oauth_token_secret: string
	}
}
```

The rest of the properties are built-in.

**Reply by deadcoder0904** [reply]
And it looks like it should work as expected:

https://github.com/nextauthjs/next-auth/blob/e8a58a01b6b8c456268771f9b382fa300ccab8ac/types/index.d.ts#L420-L427

And below `Session` extends `DefaultSession`:

https://github.com/nextauthjs/next-auth/blob/e8a58a01b6b8c456268771f9b382fa300ccab8ac/types/index.d.ts#L438

**Comment by abhaysharma3021** [comment]
Hi, I am also facing an issue. I have created a API and after login its returning id, email, and accessToken but I'm unable to set the accessToken in next-auth session as well as jwt token in next auth

**Comment by 0xBankai** [comment]
I have a similar issue.

![session_error](https://github.com/nextauthjs/next-auth/assets/49686592/bb1a21a7-13a4-45b8-b219-4becd9ec12df)


the problem is the next one:
Type '({ session }: { session: Session; }) => Promise<Session>' has no properties in common with type 'Partial<SessionOptions>'.
The expected type comes from property 'session' which is declared here on type 'AuthOptions'

**Reply by memukherjee** [reply]
The session and the signIn functions needs to be nested under callbacks.
```
const handler = NextAuth({
    providers: [
        GoogleProvider({
            clientId: process.env.GOOGLE_CLIENT_ID ?? "",
            clientSecret: process.env.GOOGLE_CLIENT_SECRET ?? "",
        }),
    ],
    callbacks: {
        async session({ session }) {
            const sessionUser = await User.findOne({
                email: session?.user?.email,
            });
            session.user.id = sessionUser?._id.toString();
            return session;
        },
        async signIn({ account, profile, user, credentials }) {
            try {
                if(!profile){
                    throw new Error("No profile found");
                }
                console.log(profile);
                
                await connectDB();

                // Check if user exists
                const userExists = await User.findOne({ email: profile.email });

                // If not, create user
                if (!userExists) {
                    await User.create({
                        email: profile.email,
                        username: profile.name?.replace(/\s/g, "").toLowerCase(),
                        image: profile.image,
                    });
                }

                return true;
            } catch (error) {
                console.log(error);
                return false;
            }
        },
    },
});
```

If you are following the the Javascript Mastery Next JS course then follow this repo:[https://github.com/adrianhajdin/project_next_13_ai_prompt_sharing/blob/main/app/api/auth/%5B...nextauth%5D/route.js](url)
![next-auth-code](https://github.com/nextauthjs/next-auth/assets/65893885/b02e9bf6-1744-4aca-8a82-982f8a72ca9e)

**Comment by develth** [comment]
How did you solved the error regarding `Awaitable`?

**Reply by diego-lipinski-de-castro** [reply]
Have you solved?

**Comment by MrOxMasTer** [comment]
typescirpt 5.5.4 / auth.js v5
all other types disappear
![image](https://github.com/user-attachments/assets/814f70fe-6db0-4bff-91c3-b70d4414553f)

**Reply by MrOxMasTer** [reply]
it helps if you put this declaration in auth.ts (didn't finish reading the documentation)
![image](https://github.com/user-attachments/assets/c948a507-f9d4-4c78-baff-8e469814e6b7)
